### DISCUSSION TITLE
How to throw Exception from server?

### DISCUSSION THREAD

**Discussion by anandsubbu007** [serverpod/serverpod#216] [Q&A] [answered]
how to throw an Exception with a status code with the message from server?

is there like ServerpodException?

**Comment by vlidholt** [comment, marked as answer]
The server will automatically set and send the HTTP status codes. 200 for successful calls, 403 if you try to access an endpoint you do not have access to or 500 for failed calls.

If you throw an exception on the server it will be logged as a failed call and a 500 code will be returned. The exception is logged to the database. Soon there will be a visual tool available to view the logs.

<img width="1259" alt="Screenshot 2022-06-03 at 16 22 13" src="https://user-images.githubusercontent.com/1539812/171873176-722db4dd-b291-451f-bfcf-2211bfd1e30a.png">

**Reply by anandsubbu007** [reply]
as per you mentioned
>  if you try to access an endpoint you do not have access

means?
is there a way to prevent users from accessing some endpoints with a scope or any other method?

**Reply by anandsubbu007** [reply]
then I can only get the error message via log only.

I think it would be great if there is a way to throw Exception  by calling methods like 
```
 throw ServerpodException(401,"Not Authorized");
```
like this.

**Reply by vlidholt** [reply]
Yes, you can set scopes on endpoints and also set the required authentication flag. There is no need to send http response codes as it is already handled by the sever automatically.

**Reply by anandsubbu007** [reply]
Great...

In some cases, I want a user to access their data only. 
if they try to alter other user's data i want to throw 401 unauthorized. 
With the help of scope, I can prevent user types like admin, and end-user. 
but I want to prevent one user to edit another user's data. 

How can I achieve this via scope?

**Reply by anandsubbu007** [reply]
https://github.com/serverpod/serverpod/pull/218#issue-1260697693

**Reply by vlidholt** [reply]
If you only want to access the data of a user, you override the `requiredLogin` property of your endpoint to force a user to be signed in. Then, you can be sure that `session.auth.authenticatedUserId` is set to the user id of the signed in user. Fetch your data and return it.

```dart
class ExampleEndpoint extends Endpoint {
  @override
  bool get requireLogin => true;

  Future<MyData> getMyData(Session session) async {
    var userId = await session.auth.authenticatedUserId;

    // Get some data with the userId

    return myData;
  }
}
```

**Reply by AndryHTC** [reply]
@vlidholt is it possible to catch **Domain Errors** thrown by the server, from client? Would be super useful to retrieve the error that caused the fail of the client call.

i.e.
## Server
```
Future createUser(String username) {
    ...
    throw UserAlreadyTaken()
    ...
}
```

## Client
```
try {
    await client.endpoint.createUser('username');
} on UserAlreadyTaken {
    // show SnackBar and suggest alternative usernames
}

```

**Reply by vlidholt** [reply]
@AndryHTC, That is a very cool idea. I cannot make it happen for the 1.0 release, but would love to make it a future feature for 1.1 or 1.2!

I added an issue here: https://github.com/serverpod/serverpod/issues/486
